src/potdata/sampler.py

```diff
@@ -206,23 +274,134 @@ def sample(self, data: list[Structure]) -> list[Structure]:
+        vectors = np.concatenate(soap_vectors, axis=1)
 
-    def _cluster(self, data: list[Structure]):
-        """Perform DBSCAN."""
+        if self.pca_dim is not None:
+            vectors = self._dim_reduction(vectors, self.pca_dim)
+
+        db = DBSCAN(eps=eps, min_samples=min_samples).fit(vectors)
+        labels = db.labels_
+
+        core_samples_mask = np.zeros_like(labels, dtype=bool)
+        core_samples_mask[db.core_sample_indices_] = True
+
+        # Number of clusters in labels, ignoring noise if present.
+        n_clusters_ = len(set(labels)) - (1 if -1 in labels else 0)
+
+        unique_labels = set(labels)
+        colors = [plt.cm.Spectral(each) for each in np.linspace(0, 1, len(unique_labels))]
+
+        # Plot noisy points and reachable points
+        noisy_points = vectors[labels == -1]
+        sampled_noisy_points = random.sample(list(noisy_points), int(noisy_ratio * len(noisy_points)))
+
+        reachable_points = vectors[np.logical_and(labels != -1, ~core_samples_mask)]
+        sampled_reachable_points = random.sample(list(reachable_points), int(reachable_ratio * len(reachable_points)))
+
+        plt.scatter(
+            [point[0] for point in sampled_noisy_points],
+            [point[1] for point in sampled_noisy_points],
+            color='black',
+            marker='o',
+            s=30,
+            label='Noisy Points'
+        )
+        plt.scatter(
+            [point[0] for point in sampled_reachable_points],
+            [point[1] for point in sampled_reachable_points],
+            color='gray',
+            marker='o',
+            s=30,
+            label='Reachable Points'
+        )
+
+        # Plot a sample of core points for each cluster
+        for k, col in zip(unique_labels, colors):
+            if k == -1:
+                continue  # Skip the noisy points
+
+            class_member_mask = labels == k
+            core_points = vectors[class_member_mask & core_samples_mask]
+            sampled_points = random.sample(list(core_points), int(core_ratio * len(core_points)))
+
+            plt.scatter(
+                [point[0] for point in sampled_points],
+                [point[1] for point in sampled_points],
+                color=tuple(col),
+                marker='o',
+                s=14,
+                label=f'Cluster {k} (Sampled)'
+            )
+
+        plt.title(f"Estimated number of clusters: {n_clusters_}")
+        plt.legend(bbox_to_anchor=(1.05, 1), loc='upper left')
+        plt.show()
+        
+        plot_clusters_with_sampling(soap_vectors)
 
-    def _compute_core_ratio(self):
+    def _compute_core_ratio(self, soap_vectors):
         """Compute the ratio of core data points to sample.
 
         ratio = min_samples/average_num_neighbors
         """
+        
+        eps = self.db_kwargs.get('eps', 0.00019)
+        min_samples = self.db_kwargs.get('min_samples', 4)
+
+        db = DBSCAN(eps=eps, min_samples=min_samples).fit(soap_vectors)
```

Comment:
Again, you should not call DBSCAN here. it should already be done in step 5. You just need to use the results there.

Reply:
Fixed